File: scripts/album_info_search.js
#!/usr/bin/env node

var  _ = require('underscore'),
  fs = require('fs'),
  os = require('os'),
  util = require('util'),
  async = require('async'),
  request = require('request'),
  readline = require('readline');

var rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

function logAlbumInfo(album) {
  console.log(
    'Album: ' + album.name + '\n' +
    'Artist: ' + album.artist
  );
}

function stepPrompt(arr, logger, callback) {
  async.detectSeries(arr,
    function (item, callback) {
      logger(item);
      rl.question('Is this correct? (y, n): ', function (res) {
        callback(/y/i.test(res));
      });
    },
    function (result) {
      rl.pause();
      if (!result) {
        callback('No more items');
      } else {
        callback(null, result);
      }
    }
  );
}

function parseAlbumSearch(body) {
  var items = JSON.parse(body).albums.slice(0, 10);
  return _.map(items, function (i) {
    i.id = i.href.split(':')[2];
    i.artist = i.artists[0].name;
    return _.pick(i, 'name', 'id', 'artist');
  });
}

function parseAlbumLookup(body) {
  var album = JSON.parse(body).album;
  album.tracks = _.map(album.tracks, function (i) {
    return i.name;
  });
  return _.pick(album, 'name', 'artist', 'released', 'tracks');
}

function genTagCmds(album) {
  return [
    genTagCmd('DATE', album.released, '*.flac'),
    genTagCmd('ALBUM', album.name, '*.flac'),
    genTagCmd('ARTIST', album.artist, '*.flac'),
    _.map(album.tracks, function (track, i) {
      var oldName;
      i = getTrackNum(i);
      oldName = genOldName(i);
      return [
        genTagCmd('TRACKNUMBER', i, oldName),
        genTagCmd('TITLE', track, oldName),
      ];
    })
  ];
}

function getTrackNum(i) {
  i += 1;
  return i < 10 ? '0' + i : i;
}

function genTagCmd(tagname, value, filename) {
  return 'metaflac --set-tag="' + tagname + '=' + value + '" ' + filename;
}

function genOldName(i) {
  return 'track' + i + '.cdda.flac';
}

function getCommandText(album) {
  var folder = '/media/external/music/iTunes/Music/"' +
    album.artist + '"/"' + album.name + '"';
  return _.flatten([
    genTagCmds(album),
    _.map(album.tracks, function (track, i) {
      i = getTrackNum(i);
      return 'mv ' + genOldName(i) + ' "' + i + ' ' + track + '.flac"';
    }),
    'mkdir -p ' + folder,
    'mv *.flac ' + folder
  ]).join(os.EOL) + os.EOL;
}


function insertNullArg(callback) {
  return function () {
    var args = _.toArray(arguments);
    args.unshift(null);
    callback.apply(this, args);
  };
}

async.waterfall([
  function (callback) {
    rl.question('Enter Album Query: ', insertNullArg(callback));
  },
  function (res, callback) {
    request('http://ws.spotify.com/search/1/album.json?q=' + res, callback);
  },
  function (res, body, callback) {
    stepPrompt(parseAlbumSearch(body), logAlbumInfo, callback);
  },
  function (res, callback) {
    request('http://ws.spotify.com/lookup/1/.json?extras=track&uri=spotify:album:' + res.id, callback);
  },
  function (res, body, callback) {
    fs.writeFile('./command', getCommandText(parseAlbumLookup(body)), callback);
  }
], function (err) {
    console.log(err || 'Done');
  }
);

File: scripts/album_info_search.py
#! /usr/bin/python3
import json
import os
import urllib.parse
import urllib.request

def writeFile(filepath, content):
  f = open(filepath, 'w')
  f.write(content)
  f.close()

def httpGet(url):
  return urllib.request.urlopen(url).readall().decode('utf-8')

def getJSON(url):
  return json.loads(httpGet(url))

def albumSearch():
  data = getJSON('http://ws.spotify.com/search/1/album.json?' +
    urllib.parse.urlencode({'q': input('Enter Album Query: ')}))
  return list(map(lambda album: {
    'name': album['name'],
    'artist': album['artists'][0]['name'],
    'id': album['href'].split(':')[2]
  }, data['albums'][0:10]))

def albumLookup(albumID):
  data = getJSON('http://ws.spotify.com/lookup/1/.json?extras=track&uri=spotify:album:' + albumID)
  tracks = []
  album = data['album']
  for i, track in enumerate(album['tracks']):
    index = format(i + 1, '02')
    tracks.append({
      'name': track['name'],
      'index': index,
      'filename': '"' + index + ' ' + track['name'] + '.flac"'
    })
  album['tracks'] = tracks
  return album

def genTagCmd(tagname, value, filename):
  return 'metaflac --set-tag="' + tagname + '=' + value + '" ' + filename

def getCommandText(album):
  folder = '~/media/music/"' +  album['artist'] + '"/"' + album['name'] + '"'
  cmds = [
    genTagCmd('DATE', album['released'], '*.flac'),
    genTagCmd('ALBUM', album['name'], '*.flac'),
    genTagCmd('ARTIST', album['artist'], '*.flac')
  ]
  for track in album['tracks']:
    cmds += [
      'mv track' + track['index'] + '.cdda.flac ' + track['filename'],
      genTagCmd('TRACKNUMBER', track['index'], track['filename']),
      genTagCmd('TITLE', track['name'], track['filename'])
    ]
  cmds += ['mkdir -p ' + folder, 'mv *.flac ' + folder]
  return os.linesep.join(cmds) + os.linesep

def selectAlbum(albums):
  selected = None
  for album in albums:
    print('Album: ' + album['name'])
    print('Artist: ' + album['artist'])
    if input('Is this correct? (y, n): ').lower() == 'y':
      selected = album
      break
  if not selected:
    print('No more items.')
  return selected

selected = selectAlbum(albumSearch())
if selected:
  album = albumLookup(selected['id'])
  writeFile('commands', getCommandText(album))
  print('Done.')
